OpenRA.Mods.Common/Activities/Air/TakeOff.cs

```diff
@@ -12,48 +12,97 @@
 using System;
 using OpenRA.Activities;
 using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
 
 namespace OpenRA.Mods.Common.Activities
 {
 	public class TakeOff : Activity
 	{
 		readonly Aircraft aircraft;
 		readonly IMove move;
-		Func<Actor, Activity, CPos, bool> moveToRallyPoint;
+		readonly Target target;
+		bool moveToRallyPoint;
+		bool soundPlayed;
 
-		public TakeOff(Actor self)
+		public TakeOff(Actor self, Target target)
 		{
 			aircraft = self.Trait<Aircraft>();
 			move = self.Trait<IMove>();
-			moveToRallyPoint = (actor, activity, pos) => NextActivity == null;
+			this.target = target;
 		}
 
-		public TakeOff(Actor self, Func<Actor, Activity, CPos, bool> moveToRallyPoint)
-			: this(self)
+		public TakeOff(Actor self)
+			: this(self, Target.FromCell(self.World, self.Location))
 		{
-			this.moveToRallyPoint = moveToRallyPoint;
+			var host = aircraft.GetActorBelow();
+			var rp = host != null ? host.TraitOrDefault<RallyPoint>() : null;
+
+			var rallyPointDestination = rp != null ? rp.Location :
+				(host != null ? self.World.Map.CellContaining(host.CenterPosition) : self.Location);
+
+			this.target = Target.FromCell(self.World, rallyPointDestination);
+			moveToRallyPoint = NextActivity == null && rallyPointDestination != self.Location;
 		}
 
 		public override Activity Tick(Actor self)
 		{
+			if (ChildActivity != null)
+			{
+				ChildActivity = ActivityUtils.RunActivity(self, ChildActivity);
+				if (ChildActivity != null)
+					return this;
+			}
+
 			// Refuse to take off if it would land immediately again.
 			if (aircraft.ForceLanding)
 			{
 				Cancel(self);
 				return NextActivity;
 			}
 
-			aircraft.UnReserve();
+			var dat = self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition);
+			if (dat <= aircraft.LandAltitude)
+			{
+				// We are taking off, so remove reservation and influence in ground cells.
```

Comment:
Shouldn't this be done from `OnFirstRun`? Otherwise we will double-tick this if we don't make it above `LandAltitude` by the second tick.